### ISSUE TITLE
Generate a suppressions.xml file from the errors list by CLI command

### ISSUE THREAD

**Issue by daniilyar** [checkstyle/checkstyle#102]
_Ported from Sorceforge [217](https://sourceforge.net/p/checkstyle/patches/217/) patch, please see all details there_

Link to patch: https://sourceforge.net/p/checkstyle/patches/_discuss/thread/1d416f58/b6be/attachment/mychanges.patch

Explanation from patch author:
http://juliangamble.com/blog/2012/03/10/running-checkstyle-at-full-throttle-on-a-legacy-codebase/

Few thoughts about reasons and usage of this patch:
http://roman-ivanov.blogspot.com/2014/01/checkstyle-suppress-file-generator.html
##

**Comment by JLLeitschuh** [comment]
Is there presently a solution to this problem?
I'm only facing 1.6k errors (down from the original 10k I had this morning) but there are still a lot of things that need to be fixed.

**Comment by rnveach** [comment]
@JLLeitschuh 
There is currently no implementation of this issue yet in Checkstyle.

An easy way to do this is to write an [`AuditListener`](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/api/AuditListener.java) that outputs the errors (when `addError` is called) to a file in the suppression format and add that to your configuration file. I don't recommend keeping it in your configuration after one run.
See: http://checkstyle.sourceforge.net/writinglisteners.html#Writing_Listeners

This is what the original poster did but the link to the source patch doesn't work anymore.

**Comment by MEZk** [comment]
@rnveach @romani 
I propose to split this issue into two parts:
1) Implementation for the old line-based suppression model as it is stable and most commonly used now;
2) Implementation for the new xpath-based suppression model;

> An easy way to do this is to write an AuditListener that outputs the errors (when addError is called) to a file in the suppression format and add that to your configuration file

No need to add the listener to the configuration file. We can add the new CLI option. If the ption is set we should add the new listener to the root module (checker) [here](https://github.com/MEZk/checkstyle/blob/2a5654734a64c5459819c5ced97bdacd0285ccc7/src/main/java/com/puppycrawl/tools/checkstyle/Main.java#L520).

In summary, CLI should have 2 options:
1) gos - generate old suppression file. 
2) gxs - generate new suppression file.

The listener should have the option to decide whether it should generate the suppression file for the old model, or for the new one.

**Comment by romani** [comment]
@MEZk ,

>Implementation for the old line-based suppression model as it is stable and most commonly used now;

It is not practical for active projects- one new line will deactivate a lot suppressions and result in a lot new to user  violations. User can do suppressions by folders/files if he need to suppress legacy code.

Please proceed only with xpath suppressions.

**Comment by MEZk** [comment]
@timurt 
Ping

**Comment by MEZk** [comment]
@timurt 
You can work on this issue.

**Comment by timurt** [comment]
@romani @rnveach 
Started working on this issue

**Comment by romani** [comment]
@timurt , I see commit with some work , but long time passed , do you plan to continue this issue ?

I have some concerns about ability to generate ready to use Xpath, as we discussed at https://github.com/checkstyle/checkstyle/pull/5762#discussion_r192256278 .

Do you plan to make generator module base on Filter interface ? exact AST is available in TreeWalkerFilters at https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/TreeWalkerAuditEvent.java#L49

**Comment by timurt** [comment]
@romani 
Yeah, I will finish this issue

Currently I faced one problem regarding tab length. `XpathQueryGenerator` requires `tab` parameter. Configuration file can store tab value as module's parameter.
Is it possible to retrieve module's tab size from `AuditListener` implementation?

**Comment by romani** [comment]
> Is it possible to retrieve module's tab size from AuditListener implementation?

no way for now.

> Or better to pass tab size as cmd parameter?

yes, http://checkstyle.sourceforge.net/cmdline.html `tabWidth`
This generator will be handy tool only, so it should not affect design of general library, in CLI we can use us much arguments as we need to make it work.

**Comment by rnveach** [comment]
> Is it possible to retrieve module's tab size from AuditListener implementation?

`AuditListener` is a checker module. It will not retrieve the AST from Treewalker filters, so I don't see how a listener will help you even if it did have the tab size.

> Or better to pass tab size as cmd parameter?

Xpaths are fully related to the Java, so this sounds like it should be a TreeWalker module of some kind as TreeWalker already has the tab size. Maybe `TreeWalkerAuditListener`? Right now, it could be hacked as a TreeWalkerFilter, but filters are suppose to suppress violations, not print outputs.

It depends on what the plans are for this. I always saw this as a listener that would either generate a file for the user based on parameters given to the module or display the generated suppressions in the output. I personally would want an options to specify if we should just suppress individual cases, suppress check for the individual file, or suppress the file.

**Comment by romani** [comment]
> Maybe TreeWalkerAuditListener? Right now, it could be hacked as a TreeWalkerFilter, but filters are suppose to suppress violations, not print outputs.

yes, lets not create new entity as TreeWalkerAuditListener for now, lets hack it as filter.

>  I personally would want an options to specify if we should just suppress individual cases, suppress check for the individual file, or suppress the file.

lets finish basic implementation and extend with fine tuning(by Check, by file,......) step by step later on (separate issues).

**Comment by timurt** [comment]
@romani 
Also question about where to store resulting xml file? 
I think I should add parameter to cmd like -osf (output suppression file), where user can specify exact path to file

Or always save file in current directory with name suppressions.xml?

**Comment by romani** [comment]
Let's skip output file parameter for now and just print result to output, user can redirect result to file himself if he needs. Most likely he will use manual merge with existing suppression file.

**Comment by romani** [comment]
after conversation with @rnveach in chat:

@timurt , you might have inconvenience to inject this custom filter to TreeWalker of user config. We are ok if you code will be hack/ugly a bit. It is ok. Filter that you might create should NOT be located in `filters` package and should NOT be documented. Please expect configs without TreeWalker, in this case we even can show some error message to user.

As we do runtime validation and xpath generation, we should be able to generate only one query, not a several as CLI do.

If you have some uncertain situations, please do not hesitate to ask.

**Comment by romani** [comment]
Fix is merged. Thanks a lot to @timurt for huge effort to make it happen.

**Comment by JLLeitschuh** [comment]
@timurt This must have been a huge task to take on, props to you for making this feature a reality!

**Comment by romani** [comment]
there in nuance in merged implementation, it was done for new filter only https://checkstyle.org/config_filters.html#SuppressionXpathFilter and this filter is more powerful BUT not compatible with all Checks for now (work is in progress).

I opened new issue #5983 to actually do this for SuppressionFilter.

Title of this issue is changed to reference SuppressionXpathFilter and do not mislead users.
